worldedit-core/src/main/java/com/sk89q/worldedit/internal/expression/Expression.java

```diff
@@ -115,9 +129,30 @@ public double evaluate(double... values) throws EvaluationException {
 
         pushInstance();
         try {
-            return root.getValue();
-        } catch (ReturnException e) {
-            return e.getValue();
+            Future<Double> result = evalThread.submit(new Callable<Double>() {
+                @Override
+                public Double call() throws Exception {
+                    return root.getValue();
+                }
+            });
+            try {
+                return result.get(WorldEdit.getInstance().getConfiguration().calculationTimeout, TimeUnit.MILLISECONDS);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
```

Comment:
I haven't worked with futures much, but doesn't this exception get thrown on the main thread?
I would think that `Future.get(...)` will interrupt the thread where the task is being run? I might be totally wrong though.

Reply:
It gets thrown if something completely unrelated to this code interrupts this current thread.

Reply:
Yes, what @sk89q said - this is just proper handling of the interrupt should we happen to get hit with one. It's an exceptional case.

Reply:
Ah ok I was not at all off. Just seemed weird how we're throwing and rethrowing runtime exceptions like candy.
Also for some reason all the expression tests are failing now. Aside from that this looks pretty good. Definitely better than the last one.